Retry reading /proc/mounts when unable to get a consistent read

jsafrane opened the pull request:
#### What type of PR is this?
/kind bug

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Helps to work around #101791

#### Special notes for your reviewer:
This is replacement of https://github.com/kubernetes/kubernetes/pull/101792

* On volume reconstruction of FC and iSCSI volume try to call `mounter.GetMountRefs()` several times (quite aggressively, 10ms delay, 1 minute!) to make sure that kubelet reconstructs a volume correctly. Any error returned from volume reconstruction may cause a volume to be marked as unused, while it's still mounted!

* Mounter tries to read /proc/mounts 10 times instead of 3 times before giving up. This will help the other volume plugins get a consistent read faster. They have other mechanism (e.g. attach) to make sure a volume cannot be mounted to different nodes, so they survive a failed reconstruction without corrupting data.

* ~WIP: this PR contains changes in kubernetes/util that needs to be merged there in a separate PR. Let me know if the idea of this PR is OK and I'll propagate the changes to kubernetes/util and re-vendor it.~

Manually tested with iSCSI and FC.

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Fixed very rare volume corruption when a pod is deleted while kubelet is offline.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

jsafrane commented:
/sig storage 

cc @jingxu97 @msau42 @gnufied, this is replacement for https://github.com/kubernetes/kubernetes/pull/102059 (fixing volume reconstruction when reading /proc/mounts fails)

gnufied commented on pkg/volume/util/hostutil/hostutil_linux.go:
```diff
@@ -111,11 +114,28 @@ func (hu *HostUtil) GetDeviceNameFromMount(mounter mount.Interface, mountPath, p
+	// Any error returned from here can result in a volume being marked as unused, while it is still
+	// mounted, therefore possibly resulting in the volume mounted elsewhere and corrupted.
+	// Try really hard (1 minute + spam the log a lot!) to get the mount refs on InconsistentReadError.
+	err := wait.PollImmediate(10*time.Millisecond, time.Minute, func() (bool, error) {
```
This function is also used for unmounting the volume in some places and hence not always for reconstruction. Unmount operation is usually retried and hence need not be blocking but reconstruction is one time operation and hence reading `/proc/mounts` is kind of blocking there (i.e if we can't read `/proc/mounts` on reconstruction - something is really weird on this host).

How about we  return `IsInconsistentReadError` error from here and then we provide another version of `hostutil.GetDeviceNameFromMount`  which has longer timeout (like 2-3 mins) and we use that version if `GetDeviceNameFromMount` returns `IsInconsistentReadError`. If we still can't read `/proc/mounts` then we should either taint the node or mark it is `NotReady`. 

I am proposing tainting the node or marking as `NotReady` because crashing the kubelet will simply cause it to go in a restart loop I think (assuming systemd or something similar is managing it). The other thing is - why we can't read `/proc/mounts` consistently? Is it because machine has high CPU load or machine has too many mount points which are coming and going? If it is later then - shouldn't we prevent other pods from being scheduled on the node? Regardless - I think we need to raise some high severity alert somewhere when this happens and one way of doing that - is to taint the node.   Having said that - if we can't read `/proc/mounts` consistently because of high CPU load or some different reason then it is not clear what would be mitigation in that case and what should user do to remove the taint or to get rid of the warning we are emitting.

jingxu97 replied:
hostutil.GetDeviceNameFromMountWIthRetry with timeout configuration?

jingxu97 replied:
also I see mount_util has a function getDeviceNameFromMount too https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/staging/src/k8s.io/mount-utils/mount.go#L190

they have different but kind of similar implementation. just wondering why we have two such functions.

jsafrane replied:
Inconsistent read from `/proc/mount` means that the mount table was modified in between two `ioutil.ReadFile()` calls. IMO it's not worth tainting the node - we try to read the file three times and expect at least two of these reads to be the same. Kubelet just had a bad luck as other processes / goroutines are mounting/unmounting other volumes.

> This function is also used for unmounting the volume in some places and hence not always for reconstruction.

It's more complicated. This function is used only for reconstruction, however, I've found GetDeviceNameFromMount copy in k8s.io/mount-utils/mount.go that's used by some drivers to unmount volumes. I'll see if they can be unified and how to add some `MustGetDeviceNameFromMount` for reconstruction.

jsafrane replied:
`hostutil.GetDeviceNameFromMount` and `mounter.GetDeviceNameFromMount` have very different semantics, the first one should be renamed to `GetRelativeGlobalVolumePath`, it returns relative dir from say `/var/lib/kubelet/plugins/kubernetes.io/aws-ebs/`. The later one returns a real device in `/dev`.

jsafrane replied:
Renamed `hostutil.GetDeviceNameFromMount` to `hostutil.GetRelativeGlobalMountPath`, which describes better what the function does.

dims commented:
/hold

jsafrane committed:
ConsistentRead tries 10 times

We've seen clusters where 3 attempts were not enough. Bumping to 10. The
slowdown should be negligible and it will reduce retry attempts in the
upper layers of kubelet.

jsafrane committed:
Retry reading /proc/mounts indifinetly in FC and iSCSI volume reconstruction

iSCSI and FC volume plugins do not implement real 3rd party attach/detach.
If reconstruction fails with an error on a FC or iSCSI volume, it will not
be unmounted from the volume global dir and at the same time it will be
marked as unused, to be available to be mounted on another node.

The volume can then be mounted on several nodes, resulting in volume
corruption.

The other block based volume plugins implement attach/detach that either
makes the volume stuck (can't be detached) or will be force-detached from a
node before attaching it somewhere else.

jsafrane commented:
I talked to @gnufied once more and we agreed that heavy retry of failed `/proc/mounts` read should be done only in iSCSI and FC volume plugins. The others have other ways (e.g. attach) to make sure a volume is not mounted twice.

I cut lot of code from the PR that was not necessary to fix the bug.

gnufied commented:
/lgtm

jsafrane commented:
Filed https://github.com/kubernetes/utils/pull/205 with the k8s.io/utils change.

jsafrane committed:
Bump k8s.io/utils

To get io/ConsistentRead updates.

jsafrane commented:
Vendored k8s.io/utils updated, now it should be merge-able

jsafrane commented:
/priority important-soon
/triage accepted

jsafrane commented:
/retest

gnufied commented:
/lgtm

jsafrane commented:
/assign @dims  
for go.mod updates and also to remove `/hold`.

dims commented:
/hold cancel
/retest
/approve

jsafrane commented:
Looks like some vendor approval magic does not work - @dims is in `dep-approvers` and this PR has only `go.mod` and related `staging/*/go.mod` and `vendor/` changes.

/assign @liggitt 
for approval. TL;DR: updating `k8s.io/utils` to the latest master to help with a bugfix.

liggitt commented on pkg/volume/util/util.go:
```diff
@@ -732,3 +734,29 @@ func IsDeviceMountableVolume(volumeSpec *volume.Spec, volumePluginMgr *volume.Vo
+		paths, err = mounter.GetMountRefs(mountPath)
+		if io.IsInconsistentReadError(err) {
+			klog.Errorf("Failed to read mount refs from /proc/mounts: %s", mountPath, err)
+			klog.Errorf("The system looks heavily overloaded, will retry indefinitely. Kubelet cannot unmount volumes without getting consistent /proc/mounts")
```
"indefinitely" up to a minute?

jsafrane replied:
There was endless loop initially, with extra spammy logs. Now that it tries only for one minute it makes sense to log the error only at the end.

liggitt commented on pkg/volume/util/util.go:
```diff
@@ -732,3 +734,29 @@ func IsDeviceMountableVolume(volumeSpec *volume.Spec, volumePluginMgr *volume.Vo
+		var err error
+		paths, err = mounter.GetMountRefs(mountPath)
+		if io.IsInconsistentReadError(err) {
+			klog.Errorf("Failed to read mount refs from /proc/mounts: %s", mountPath, err)
+			klog.Errorf("The system looks heavily overloaded, will retry indefinitely. Kubelet cannot unmount volumes without getting consistent /proc/mounts")
```
these messages will log every ~10ms up to a minute (6000 times)?

jsafrane replied:
log removed

liggitt commented on pkg/volume/util/util.go:
```diff
@@ -732,3 +734,29 @@ func IsDeviceMountableVolume(volumeSpec *volume.Spec, volumePluginMgr *volume.Vo
+			return false, err
+		}
+		if errLogged {
+			klog.Infof("Got consistent read of /proc/mounts")
```
does this need to include the mount path to differentiate from other concurrent errors?

jsafrane replied:
Reworked the logging completely, now it always logs the affected path.

liggitt commented on staging/src/k8s.io/mount-utils/mount_helper_unix.go:
```diff
@@ -32,7 +32,7 @@ const (
 	expectedAtLeastNumFieldsPerMountInfo = 10
 	// How many times to retry for a consistent read of /proc/mounts.
-	maxListTries = 3
+	maxListTries = 10
```
what are the performance implications of tripling this?

jsafrane replied:
Almost none, I was able to get tens of thousands of /proc/mounts reads per second. So... under 1 ms?

liggitt commented on pkg/volume/util/util.go:
```diff
@@ -732,3 +734,29 @@ func IsDeviceMountableVolume(volumeSpec *volume.Spec, volumePluginMgr *volume.Vo
+		return true, nil
+	})
+	if err == wait.ErrWaitTimeout {
+		klog.Errorf("Failed to read mount refs from /proc/mounts for %s: %s", mountPath, err)
```
this will log a timeout error, not the inconsistent read error... is that what you wanted?

jsafrane replied:
Fixed (by moving the reporting to the caller)

liggitt commented on pkg/volume/util/util.go:
```diff
@@ -732,3 +734,29 @@ func IsDeviceMountableVolume(volumeSpec *volume.Spec, volumePluginMgr *volume.Vo
+	})
+	if err == wait.ErrWaitTimeout {
+		klog.Errorf("Failed to read mount refs from /proc/mounts for %s: %s", mountPath, err)
+		klog.Errorf("Kubelet cannot unmount volume at %s, please unmount it manually", mountPath)
```
this seems like an odd message to log from a method that is only reading mount refs... shouldn't this be logged by callers that are blocked from unmounting because of this?

jsafrane replied:
Ok, I moved the error reporting to the caller, but it's copied to two places now.

liggitt replied:
still need to push?

jsafrane committed:
Move error reporting to volume plugins

Move reporting of GetReliableMountRefs error to the volume plugins that
have more context about severity of the error.

jsafrane commented:
@liggitt can you please take a look again?

jsafrane commented:
sorry, pushed right now.

jsafrane commented:
@liggitt, can you please check the PR again?

liggitt commented:
/assign @msau42 

would like a second set of eyes from storage. dep update looks fine, will tag once msau takes a look

msau42 commented:
/lgtm
This logic only impacts fc and iscsi plugins during reconstruction, which only happens once on kubelet startup.

msau42 commented:
/retest

liggitt commented:
/approve
